File: src/main/java/com/github/windchopper/tools/eq2/lp/EventBrowserStageController.java
package com.github.windchopper.tools.eq2.lp;

import com.github.windchopper.common.fx.application.fx.StageController;
import com.github.windchopper.common.fx.application.fx.annotation.FXMLResource;
import com.github.windchopper.common.fx.application.fx.event.FXMLResourceOpen;
import com.github.windchopper.common.util.KnownSystemProperties;
import com.github.windchopper.common.util.Pipeliner;
import com.github.windchopper.tools.eq2.lp.log.LogRecord;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.control.TreeTableView;
import javafx.stage.FileChooser;
import javafx.stage.Modality;
import javafx.stage.Stage;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Event;
import javax.inject.Inject;
import java.io.File;
import java.util.HashMap;
import java.util.Optional;

@ApplicationScoped @FXMLResource(Globals.FXMLResources.FXML__EVENT_BROWSER_STAGE) public class EventBrowserStageController
    extends StageController implements KnownSystemProperties {

    @Inject private Event<FXMLResourceOpen> fxmlResourceOpenEvent;
    @FXML private TreeTableView<LogRecord> eventTreeTableView;

    @FXML protected void openFile(ActionEvent event) {
        Optional.ofNullable(Pipeliner.of(FileChooser::new)
            .set(target -> target::setInitialDirectory, Optional.ofNullable(Globals.Settings.openFileInitialDirectory.get())
                .orElse(userHomeFile.get().orElse(
                    new File(""))))
            .map(fileChooser -> fileChooser.showOpenMultipleDialog(stage))
            .get()).ifPresent(files -> {
                files.stream().findAny().map(File::getParentFile).ifPresent(Globals.Settings.openFileInitialDirectory);
                fxmlResourceOpenEvent.fire(
                    new FXMLResourceOpen(
                        Pipeliner.of(Stage::new)
                            .set(target -> target::initOwner, stage)
                            .set(target -> target::initModality, Modality.APPLICATION_MODAL)
                            .set(target -> target::setResizable, false)
                            .get(),
                        Globals.FXMLResources.FXML__SELECT_LOG_EVENT_BUILDER_STAGE,
                        Pipeliner.of(HashMap<String, Object>::new)
                            .accept(target -> target.put("files", files))
                            .get()));
            });
    }

}

File: src/main/java/com/github/windchopper/tools/eq/lp/EventBrowserStageController.kt
package com.github.windchopper.tools.eq.lp

import com.github.windchopper.common.fx.cdi.form.Form
import com.github.windchopper.common.fx.cdi.form.StageFormController
import com.github.windchopper.common.fx.cdi.form.StageFormLoad
import com.github.windchopper.common.util.ClassPathResource
import com.github.windchopper.common.util.Pipeliner
import com.github.windchopper.tools.eq.lp.log.LogRecord
import jakarta.enterprise.context.ApplicationScoped
import jakarta.enterprise.event.Event
import jakarta.inject.Inject
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.scene.control.TreeTableView
import javafx.stage.FileChooser
import javafx.stage.Modality
import javafx.stage.Stage
import java.io.File
import java.util.*
import java.util.function.Consumer

@ApplicationScoped
@Form(Application.FXML__EVENT_BROWSER_STAGE)
class EventBrowserStageController: StageFormController() {
    @Inject
    private val fxmlResourceOpenEvent: Event<StageFormLoad>? = null

    @FXML
    private val eventTreeTableView: TreeTableView<LogRecord>? = null

    @FXML
    fun openFile(event: ActionEvent?) {
        val fileChooser = Pipeliner.of { FileChooser() }
            .set({ target: FileChooser -> Consumer { file: File? -> target.initialDirectory = file } }, Optional.ofNullable(Application.openFileInitialDirectory.load())
                .orElseGet {
                    val initialDirectory = File(System.getProperty("user.home"))
                    Application.openFileInitialDirectory.save(initialDirectory)
                    initialDirectory
                })
            .get()
        Optional.ofNullable(fileChooser.showOpenMultipleDialog(stage))
            .ifPresent { files: List<File>? ->
                fxmlResourceOpenEvent!!.fire(
                    StageFormLoad(
                        ClassPathResource(Application.FXML__SELECT_LOG_EVENT_BUILDER_STAGE),
                        mapOf("files" to files),
                        Pipeliner.of { Stage() }
                            .set({ target: Stage -> Consumer { window: Stage? -> target.initOwner(window) } }, stage)
                            .set({ target: Stage -> Consumer { modality: Modality? -> target.initModality(modality) } }, Modality.APPLICATION_MODAL)
                            .set({ target: Stage -> Consumer { b: Boolean? -> target.isResizable = b!! } }, false)))
            }
    }
}